reduct/enf/Utilities/parser.py

```diff
@@ -0,0 +1,78 @@
+import re
+
+## accepts Metta boolean expression and returns equivalent binary enf expression
+## e.g 
+##      (OR A B C D) -->  "|(a,|(b,|(c,d)))
+def parse_metta_expression(expr: str) -> str:
+    def parse_tokens(tokens: list[str]) -> str:
+        token = tokens.pop(0)
+
+        if token != '(':
+            return token.lower()
+
+        op = tokens.pop(0).upper()
+
+        if op == 'NOT':
+            sub_expr = parse_tokens(tokens)
+            assert tokens.pop(0) == ')'
+            return f"!({sub_expr})"
+
+        elif op in ('AND', 'OR'):
+            symbol = '&' if op == 'AND' else '|'
+            args = []
+            while tokens[0] != ')':
+                args.append(parse_tokens(tokens))
+            tokens.pop(0)  # remove ')'
+
+            # Convert to binary expression by right-associating
+            def to_binary(args):
+                if len(args) == 1:
+                    return args[0]
+                return f"{symbol}({args[0]},{to_binary(args[1:])})"
+            
+            return to_binary(args)
+
+        else:
+            raise ValueError(f"Unknown operator: {op}")
+
+    tokens = re.findall(r'\(|\)|[A-Za-z]+', expr)
+    return parse_tokens(tokens)
+
+## accepts constraint tree and returns metta boolean expression
```

Comment:
Making it clear in the comment that the constraint tree mentioned here is a constraint tree's string representation would be a lot helpful.